### PR TITLE
Add scrollable table component WD-3408

### PR DESCRIPTION
**Pull request by edlerd**
## Done

- added reusable scrollable table component

Fixes WD-3408

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described in the [Readme](https://github.com/canonical/lxd-ui#setting-up-for-development).
2. Perform the following QA steps:
    - check scrollable tables  (instance list, snapshot list, profile list, profile/instance configuration)

### CODE REVIEW

**Comment by lorumic** [comment]
A very high-level QA comment before diving into the code. The mechanism to auto-hide columns on medium screens in the instance table seems to be broken. See the following comparison with 1240px width:

BEFORE:
image

AFTER:
image

**Comment by edlerd** [comment]
> A very high-level QA comment before diving into the code. The mechanism to auto-hide columns on medium screens in the instance table seems to be broken.

That is fixed just now. I had in the back of my head the idea to use a ref instead of giving a static `id` and using `getElementById`, which turns out to fix this issue as well :)

File: src/pages/instances/InstanceList.tsx [hunk 1]
```diff
@@ -535,30 +521,42 @@ const InstanceList: FC = () => {
...
+                      instances,
+                      creationOperations,
+                      notify.notification,
+                      filters.queries,
+                      filters.statuses,
+                      filters.types,
+                      filters.profileQueries,
+                      pagination.pageSize,
+                      pagination.currentPage,
+                    ]}
```
Comment (lorumic) [inline src/pages/instances/InstanceList.tsx:535]: This looks quite chaotic. Can we create a `const` for this array up there and reference it here?

File: src/pages/instances/InstanceSnapshots.tsx [hunk 2]
```diff
@@ -258,24 +250,32 @@ const InstanceSnapshots: FC<Props> = ({ instance }) => {
...
-            defaultSortDirection="descending"
-          />
+          <ScrollableTable
+            dependencies={[
+              instance.snapshots,
+              inTabNotification,
+              query,
+              pagination.pageSize,
+              pagination.currentPage,
+            ]}
```
Comment (lorumic) [inline src/pages/instances/InstanceSnapshots.tsx:260]: Here too, and in all the other tables, I'd move this dependency array up in a `const` to make it easier to maintain.

File: src/components/ScrollableTable.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const ref = useRef<HTMLDivElement>(null);
+
+  const updateTBodyHeight = () => {
+    const table = ref.current?.children[0];
+    if (!table || table.children.length !== 2) {
+      return;
+    }
+    const tBody = table.children[1];
+    const above = tBody.getBoundingClientRect().top + 1;
+    const below = 85;
```
[the remark this answers is not in the document]
Comment (edlerd) [reply src/components/ScrollableTable.tsx]: True, I changed this by adding an optional `belowId` as a string of which we now read height + margintop + marginbottom to figure the right value.

File: src/pages/instances/InstanceList.tsx (hunk 1 above)
Comment (edlerd) [reply src/pages/instances/InstanceList.tsx]: I think I found a simplification to pass the notification and the filtered object list. The rest is not needed.

File: src/pages/instances/InstanceSnapshots.tsx (hunk 2 above)
Comment (edlerd) [reply src/pages/instances/InstanceSnapshots.tsx]: same as above, shortened the list.

File: src/components/ScrollableTable.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (lorumic) [reply src/components/ScrollableTable.tsx]: And you know what's the best about this? That incidentally, this fixes one of the 2 issues with the search and filter:

Peek 2023-05-25 23-27

:tada: :clap:

**Review by lorumic — APPROVED** [review]

**Pull request merged**